skypy/galaxy/redshift.py

```diff
@@ -5,243 +5,216 @@
+    noise. If `noise` is false, the expected number of galaxies is used.
 
-     References
+    Parameters
     ----------
-    .. [1] Herbel J., Kacprzak T., Amara A. et al., 2017, Journal of Cosmology
-           and Astroparticle Physics, Issue 08, article id. 035 (2017)
+    redshift : array_like
+        Redshifts at which comoving number densities are provided.
+    density : array_like
+        Comoving galaxy number number density at each redshift in Mpc-3.
+    fsky : array_like
+        Sky fraction over which galaxies are sampled.
+    cosmology : Cosmology, optional
+        Cosmology object for conversion to comoving volume. If not given, the
+        default cosmology is used.
+    noise : bool, optional
+        Poisson-sample the number of galaxies. Default is `True`.
+
+    Returns
+    -------
+    redshifts : array_like
+        Sampled redshifts such that the comoving number density of galaxies
+        corresponds to the input distribution.
 
     Examples
     --------
-    >>> from skypy.galaxy.redshift import herbel_redshift
-    >>> from astropy.cosmology import FlatLambdaCDM
+    Sample redshifts with a constant comoving number density 1e-3/Mpc3 up to
+    redshift 1 for a survey of 1 square degree = 1/41253 of the sky.
 
-    Sample 100 redshift values from the Schechter luminosity function with
-    a_m = -0.9408582, b_m = -20.40492365, a_phi = -0.10268436,
-    b_phi = 0.00370253, alpha = -1.3.
+    >>> from skypy.galaxy.redshift import redshifts_from_comoving_density
+    >>> z_range = np.arange(0, 1.01, 0.1)
+    >>> z_gal = redshifts_from_comoving_density(z_range, 1e-3, 1/41253)
 
-    >>> cosmology = FlatLambdaCDM(H0=70, Om0=0.3, Tcmb0=2.725)
-    >>> redshift = herbel_redshift(size=1000, low=0.01, alpha=-1.3,
-    ...                     a_phi=-0.10268436,a_m=-0.9408582, b_phi=0.00370253,
-    ...                     b_m=-20.40492365, cosmology=cosmology,
-    ...                     absolute_magnitude_max=-16.)
+    '''
 
-    """
+    # redshift number density
+    dN_dz = cosmology.differential_comoving_volume(redshift).to_value('Mpc3/sr')
+    dN_dz *= density
+    dN_dz *= 4*np.pi*fsky
 
-    redshift = np.linspace(low, high, resolution)
-    pdf = herbel_pdf(redshift, alpha, a_phi, b_phi, a_m, b_m, cosmology,
-                     absolute_magnitude_max)
-    cdf = scipy.integrate.cumtrapz(pdf, redshift, initial=0)
-    cdf = cdf / cdf[-1]
-    u = np.random.uniform(size=size)
-    redshift_sample = np.interp(u, cdf, redshift)
+    # integrate density to get expected number of galaxies
+    N = np.trapz(dN_dz, redshift)
 
-    return redshift_sample
+    # Poisson sample galaxy number if requested
+    if noise:
+        N = np.random.poisson(N)
+    else:
+        N = int(N)
 
+    # cumulative trapezoidal rule to get redshift CDF
+    cdf = dN_dz  # reuse memory
+    np.cumsum((dN_dz[1:]+dN_dz[:-1])/2*np.diff(redshift), out=cdf[1:])
```

Comment:
There is no numpy cumulative trapezoidal rule, and I found it over the top to import scipy.integrate for using the same one-liner. I agree that it would be better to have an automatic integration but it's more tricky than one might think, so I would need to think a little about that

Reply:
Actually no, in this function you only have the density given at a fixed resolution, so there's not much else you can do.

Reply:
And using scipy is not a good alternative?